mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -116,18 +118,110 @@ class AdjointGenerator {
                 auto clone = cast<QuantumGate>(builder.clone(*gate, remappedValues));
                 clone.setAdjointFlag(!gate.getAdjointFlag());
 
-                // Read cached differentiable parameters from the recorded parameter vector.
-                if (auto differentiableGate = dyn_cast<quantum::DifferentiableGate>(op)) {
+                // Read cached parameters from the recorded parameter vector.
+                if (auto parametrizedGate = dyn_cast<quantum::ParametrizedGate>(op)) {
                     OpBuilder::InsertionGuard insertionGuard(builder);
                     builder.setInsertionPoint(clone);
                     SmallVector<Value> cachedParams;
-                    ValueRange diffParams = differentiableGate.getDiffParams();
-                    for (unsigned i = 0; i < diffParams.size(); i++) {
-                        cachedParams.push_back(builder.create<ListPopOp>(
-                            differentiableGate.getLoc(), cache.paramVector));
+                    ValueRange params = parametrizedGate.getAllParams();
+                    for (Value param : params) {
```

Comment:
Same thing about raising an error vs assertion in this piece :)

Reply:
The reason why I preferred assertions here is because we are guaranteed by the verifier that all of these properties will hold. It is not until the code is extended that we need to check the invariants again. I have added the errors and now CodeFactor complains because there are too many if statements :sweat_smile:

Reply:
> The reason why I preferred assertions here is because we are guaranteed by the verifier that all of these properties will hold.

The problem is that verifier of the type you are matching against (ParametrizedGate) contains no such guarantees. It is only with the implicit additional assumption "only `CustomOp`, `QubitUnitary`, and `MultiRZ` are `ParametrizedGate`s" that the guarantee is given. Something that can easily be broken in the future.

Generally assertions should be fine, but given that they can easily be disabled it is not the same thing as error checks.

> I have added the errors and now CodeFactor complains because there are too many if statements 😅

That is annoying indeed 😅